openmp/libomptarget/plugins/opencl/src/rtl.cpp

```diff
@@ -0,0 +1,3625 @@
+        DeviceInfo->getExtensionFunctionName(
+            DeviceId, clSetProgramSpecializationConstantId));
+    return;
+  }
+
+  for (int I = ConstantValues.size(); I > 0; --I) {
+    cl_uint Id = static_cast<cl_uint>(ConstantIds[I - 1]);
+    size_t Size = ConstantValueSizes[I - 1];
+    const void *Val = ConstantValues[I - 1];
+    CALL_CL_EXT_SILENT(DeviceId, Rc, clSetProgramSpecializationConstant,
+                       Program, Id, Size, Val);
+    if (Rc == CL_SUCCESS)
+      IDP("Set specialization constant '0x%X'\n", static_cast<int32_t>(Id));
+  }
+}
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+static inline void addDataTransferLatency() {
+  if (!DeviceInfo->Flags.CollectDataTransferLatency)
+    return;
+  double goal = omp_get_wtime() + 1e-6 * DeviceInfo->DataTransferLatency;
+  // Naive spinning should be enough
+  while (omp_get_wtime() < goal)
+    ;
+}
+
+int32_t ExtensionsTy::getExtensionsInfoForDevice(int32_t DeviceNum) {
+  // Identify the size of OpenCL extensions string.
+  size_t RetSize = 0;
+  // If the below call fails, some extensions's status may be
+  // left ExtensionStatusUnknown, so only ExtensionStatusEnabled
+  // actually means that the extension is enabled.
+  IDP("Getting extensions for device %d\n", DeviceNum);
+
+  cl_device_id DeviceId = DeviceInfo->deviceIDs[DeviceNum];
+  CALL_CL_RET_FAIL(clGetDeviceInfo, DeviceId, CL_DEVICE_EXTENSIONS, 0, nullptr,
+                   &RetSize);
+
+  std::unique_ptr<char []> Data(new char[RetSize]);
+  CALL_CL_RET_FAIL(clGetDeviceInfo, DeviceId, CL_DEVICE_EXTENSIONS, RetSize,
+                   Data.get(), &RetSize);
+
+  std::string Extensions(Data.get());
+  IDP("Device extensions: %s\n", Extensions.c_str());
+
+  if (UnifiedSharedMemory == ExtensionStatusUnknown &&
+      Extensions.find("cl_intel_unified_shared_memory") != std::string::npos) {
+    UnifiedSharedMemory = ExtensionStatusEnabled;
+    IDP("Extension UnifiedSharedMemory enabled.\n");
+  }
+
+  std::for_each(LibdeviceExtensions.begin(), LibdeviceExtensions.end(),
+                [&Extensions](LibdeviceExtDescTy &Desc) {
+                  if (Desc.Status == ExtensionStatusUnknown)
+                    if (Extensions.find(Desc.Name) != std::string::npos) {
+                      Desc.Status = ExtensionStatusEnabled;
+                      IDP("Extension %s enabled.\n", Desc.Name);
+                    }
+                });
+
+  return CL_SUCCESS;
+}
+
+// FIXME: move this to llvm/BinaryFormat/ELF.h and elf.h:
+#define NT_INTEL_ONEOMP_OFFLOAD_VERSION 1
+#define NT_INTEL_ONEOMP_OFFLOAD_IMAGE_COUNT 2
+#define NT_INTEL_ONEOMP_OFFLOAD_IMAGE_AUX 3
+
+static bool isValidOneOmpImage(__tgt_device_image *Image,
+                               uint64_t &MajorVer,
+                               uint64_t &MinorVer) {
+  char *ImgBegin = reinterpret_cast<char *>(Image->ImageStart);
+  char *ImgEnd = reinterpret_cast<char *>(Image->ImageEnd);
+  size_t ImgSize = ImgEnd - ImgBegin;
+  ElfL E(ImgBegin, ImgSize);
```

Comment:
Does this change-set have to be buildable?  `ElfL` is defined in `plugins/common`.

Reply:
@vzakhari , the added files are not intended to be part of the build. The CMakeLists.txt is included for reference only. I'm relying on the fact that the parent-level openmp/libomptarget/plugins/CMakeLists.txt does not say "add_subdirectory(opencl)"; does that guarantee that the build will ignore stuff under plugins/opencl/ ?